File: frontend/apps/ios/Pantopus/Features/Feed/Pulse/PulseIntent.swift
//
//  PulseIntent.swift
//  Pantopus
//
//  Six-way classification for Pulse posts. Drives the chip-row filter,
//  the per-card colored chip, the reaction-verb set, and the compose
//  FAB's pre-fill. The `all` case is a chip-row-only sentinel; real
//  posts always resolve to one of the other five.
//

import Foundation

/// One of the six chip-row intents.
public enum PulseIntent: String, CaseIterable, Sendable, Hashable {
    case all
    case ask
    case recommend
    case event
    case lost
    case announce

    /// Chip-row display label.
    public var label: String {
        switch self {
        case .all: "All"
        case .ask: "Ask"
        case .recommend: "Recommend"
        case .event: "Event"
        case .lost: "Lost & Found"
        case .announce: "Announce"
        }
    }

    /// Right-aligned per-card chip label (shorter than the chip-row label).
    public var cardChipLabel: String {
        switch self {
        case .all: ""
        case .ask: "Ask"
        case .recommend: "Rec"
        case .event: "Event"
        case .lost: "Lost"
        case .announce: "Announce"
        }
    }

    /// Backend `post_type` filter value sent on `/api/posts/feed`. `all`
    /// returns `nil` so the backend skips the filter.
    public var postType: String? {
        switch self {
        case .all: nil
        case .ask: "ask_local"
        case .recommend: "recommendation"
        case .event: "event"
        case .lost: "lost_found"
        case .announce: "local_update"
        }
    }

    /// Resolve a backend `post_type` string back to a UI intent. Unknown
    /// types fall through to `.announce` — the most generic chip — so
    /// the card still renders a meaningful intent indicator.
    public static func from(postType: String?) -> PulseIntent {
        switch postType ?? "" {
        case "ask_local", "ask": .ask
        case "recommendation", "recommend": .recommend
        case "event": .event
        case "lost_found": .lost
        case "local_update", "announcement", "heads_up", "neighborhood_win", "alert": .announce
        default: .announce
        }
    }
}

public extension PulseIntent {
    /// Pantopus icon used inside the per-card intent chip.
    var icon: PantopusIcon {
        switch self {
        case .all: .info
        case .ask: .helpCircle
        case .recommend: .thumbsUp
        case .event: .calendar
        case .lost: .search
        case .announce: .megaphone
        }
    }
}

/// One reaction kind shown in the bottom strip of a post card. The
/// backend only persists `like` (helpful); the other counts are
/// display-only and intent-shaped to match the design.
public struct PulseReaction: Sendable, Hashable, Identifiable {
    public enum Kind: String, Sendable, Hashable {
        case helpful, heart, going, seen, shared
    }

    public let id: Kind
    public let kind: Kind
    public let icon: PantopusIcon
    public let label: String
    public let count: Int
    /// True iff this reaction maps to the `like` toggle endpoint.
    public let isInteractive: Bool

    public init(kind: Kind, icon: PantopusIcon, label: String, count: Int, isInteractive: Bool) {
        id = kind
        self.kind = kind
        self.icon = icon
        self.label = label
        self.count = count
        self.isInteractive = isInteractive
    }
}

public extension PulseIntent {
    /// Returns the reaction strip the design specifies for this intent.
    /// The first kind is always the one wired to `POST /:id/like`; the
    /// rest are display-only counts.
    func reactionTemplate(helpfulCount: Int, secondaryCount: Int = 0) -> [PulseReaction] {
        switch self {
        case .ask:
            [
                PulseReaction(kind: .helpful, icon: .lightbulb, label: "helpful", count: helpfulCount, isInteractive: true),
                PulseReaction(kind: .heart, icon: .heart, label: "", count: secondaryCount, isInteractive: false)
            ]
        case .recommend:
            [
                PulseReaction(kind: .helpful, icon: .heart, label: "", count: helpfulCount, isInteractive: true),
                PulseReaction(kind: .heart, icon: .lightbulb, label: "helpful", count: secondaryCount, isInteractive: false)
            ]
        case .event:
            [
                PulseReaction(kind: .going, icon: .calendarCheck, label: "going", count: helpfulCount, isInteractive: true),
                PulseReaction(kind: .heart, icon: .heart, label: "", count: secondaryCount, isInteractive: false)
            ]
        case .lost:
            [
                PulseReaction(kind: .seen, icon: .eye, label: "seen", count: helpfulCount, isInteractive: true),
                PulseReaction(kind: .shared, icon: .share, label: "shared", count: secondaryCount, isInteractive: false)
            ]
        case .announce, .all:
            [
                PulseReaction(kind: .helpful, icon: .lightbulb, label: "helpful", count: helpfulCount, isInteractive: true),
                PulseReaction(kind: .heart, icon: .heart, label: "", count: secondaryCount, isInteractive: false)
            ]
        }
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/feed/pulse/PulseIntent.kt
@file:Suppress("MagicNumber")

package app.pantopus.android.ui.screens.feed.pulse

import androidx.compose.runtime.Immutable
import app.pantopus.android.ui.theme.PantopusIcon

/**
 * Six-way classification for Pulse posts. Drives the chip-row filter,
 * the per-card colored chip, the reaction-verb set, and the compose
 * FAB's pre-fill. `All` is a chip-row-only sentinel; real posts always
 * resolve to one of the other five.
 */
enum class PulseIntent(
    val key: String,
    val label: String,
    val cardChipLabel: String,
) {
    All(key = "all", label = "All", cardChipLabel = ""),
    Ask(key = "ask", label = "Ask", cardChipLabel = "Ask"),
    Recommend(key = "recommend", label = "Recommend", cardChipLabel = "Rec"),
    Event(key = "event", label = "Event", cardChipLabel = "Event"),
    Lost(key = "lost", label = "Lost & Found", cardChipLabel = "Lost"),
    Announce(key = "announce", label = "Announce", cardChipLabel = "Announce"),
    ;

    /** Backend `post_type` filter for `/api/posts/feed`. `All` is `null`. */
    val postType: String?
        get() =
            when (this) {
                All -> null
                Ask -> "ask_local"
                Recommend -> "recommendation"
                Event -> "event"
                Lost -> "lost_found"
                Announce -> "local_update"
            }

    /** Icon used inside the per-card intent chip. */
    val icon: PantopusIcon
        get() =
            when (this) {
                All -> PantopusIcon.Info
                Ask -> PantopusIcon.HelpCircle
                Recommend -> PantopusIcon.ThumbsUp
                Event -> PantopusIcon.Calendar
                Lost -> PantopusIcon.Search
                Announce -> PantopusIcon.Megaphone
            }

    companion object {
        fun fromKey(key: String): PulseIntent = entries.firstOrNull { it.key == key } ?: All

        /**
         * Resolve a backend `post_type` to a UI intent. Unknown values
         * fall through to `Announce` (most generic) so the card still
         * renders a meaningful indicator.
         */
        fun fromPostType(postType: String?): PulseIntent =
            when (postType ?: "") {
                "ask_local", "ask" -> Ask
                "recommendation", "recommend" -> Recommend
                "event" -> Event
                "lost_found" -> Lost
                "local_update", "announcement", "heads_up", "neighborhood_win" -> Announce
                else -> Announce
            }
    }
}

/**
 * One reaction kind shown in the bottom strip of a post card. The
 * backend only persists `like` (helpful); the other counts are
 * display-only and intent-shaped to match the design.
 */
@Immutable
data class PulseReaction(
    val kind: Kind,
    val icon: PantopusIcon,
    val label: String,
    val count: Int,
    val isInteractive: Boolean,
) {
    enum class Kind(val key: String) {
        Helpful("helpful"),
        Heart("heart"),
        Going("going"),
        Seen("seen"),
        Shared("shared"),
    }
}

/**
 * Returns the reaction strip the design specifies for this intent.
 * The first kind is wired to `POST /:id/like`; the rest are
 * display-only counts.
 */
fun PulseIntent.reactionTemplate(
    helpfulCount: Int,
    secondaryCount: Int = 0,
): List<PulseReaction> =
    when (this) {
        PulseIntent.Ask ->
            listOf(
                PulseReaction(PulseReaction.Kind.Helpful, PantopusIcon.Lightbulb, "helpful", helpfulCount, true),
                PulseReaction(PulseReaction.Kind.Heart, PantopusIcon.Heart, "", secondaryCount, false),
            )
        PulseIntent.Recommend ->
            listOf(
                PulseReaction(PulseReaction.Kind.Helpful, PantopusIcon.Heart, "", helpfulCount, true),
                PulseReaction(PulseReaction.Kind.Heart, PantopusIcon.Lightbulb, "helpful", secondaryCount, false),
            )
        PulseIntent.Event ->
            listOf(
                PulseReaction(PulseReaction.Kind.Going, PantopusIcon.Check, "going", helpfulCount, true),
                PulseReaction(PulseReaction.Kind.Heart, PantopusIcon.Heart, "", secondaryCount, false),
            )
        PulseIntent.Lost ->
            listOf(
                PulseReaction(PulseReaction.Kind.Seen, PantopusIcon.Eye, "seen", helpfulCount, true),
                PulseReaction(PulseReaction.Kind.Shared, PantopusIcon.Share, "shared", secondaryCount, false),
            )
        PulseIntent.Announce, PulseIntent.All ->
            listOf(
                PulseReaction(PulseReaction.Kind.Helpful, PantopusIcon.Lightbulb, "helpful", helpfulCount, true),
                PulseReaction(PulseReaction.Kind.Heart, PantopusIcon.Heart, "", secondaryCount, false),
            )
    }
